src/forwarder.rs

```diff
@@ -0,0 +1,394 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Connecting => write!(f, "Connecting"),
+            Self::Connected => write!(f, "Connected"),
+            Self::Disconnected => write!(f, "Disconnected"),
+        }
+    }
+}
+
+struct SessionState {
+    token: String,
+    status: SessionStatus,
+}
+
+/// Struct representing the state of a remote session with a device
+impl SessionState {
+    fn connecting(token: String) -> Self {
+        Self {
+            token,
+            status: SessionStatus::Connecting,
+        }
+    }
+
+    fn connected(token: String) -> Self {
+        Self {
+            token,
+            status: SessionStatus::Connected,
+        }
+    }
+
+    fn disconnected(token: String) -> Self {
+        Self {
+            token,
+            status: SessionStatus::Disconnected,
+        }
+    }
+}
+
+impl From<SessionState> for AstarteType {
+    fn from(value: SessionState) -> Self {
+        match value.status {
+            SessionStatus::Connecting | SessionStatus::Connected => {
+                Self::String(value.status.to_string())
+            }
+            SessionStatus::Disconnected => Self::Unset,
+        }
+    }
+}
+
+/// Device forwarder.
+///
+/// It maintains a collection of tokio task handles, each one identified by a [`Key`] containing
+/// the connection information and responsible for providing forwarder functionalities. For
+/// instance, a task could open a remote terminal between the device and a certain host.
+#[derive(Debug)]
+pub struct Forwarder {
+    tx_state: Sender<SessionState>,
+    tasks: HashMap<Key, JoinHandle<()>>,
+}
+
+impl Forwarder {
+    /// Initialize the forwarder instance, spawning also a task responsible for sending a property
+    /// necessary to update the device session state.
+    pub fn init<P>(publisher: P) -> Self
+    where
+        P: Publisher + 'static + Send + Sync,
+    {
+        let (tx_state, rx_state) = channel::<SessionState>(CHANNEL_STATE_SIZE);
+
+        // the handle is not stored because it is ended when all tx_state are dropped (which is when
+        // the Forwarder instance is dropped
+        tokio::spawn(Self::handle_session_state_task(publisher, rx_state));
+
+        Self {
+            tx_state,
+            tasks: Default::default(),
+        }
+    }
+
+    async fn handle_session_state_task<P>(publisher: P, mut rx_state: Receiver<SessionState>)
+    where
+        P: Publisher,
+    {
+        while let Some(msg) = rx_state.recv().await {
+            let ipath = format!("/{}/status", msg.token);
+            let idata = msg.into();
+
+            if let Err(err) = publisher
+                .send(
+                    "io.edgehog.devicemanager.ForwarderSessionState",
+                    &ipath,
+                    idata,
+                )
+                .await
+            {
+                error!("publisher send error, {err}");
+            }
+        }
+    }
```

Comment:
This can be handle in the connection task

Reply:
by spawning it in the `init` function the channel is being defined only, independently from the opened session. 
Btw, where do you mean exactly?